backend/cdApp.py

```diff
@@ -37,35 +37,138 @@ def dict_factory(cursor, row):
     return d
+
+#---------------------HTTP Errors---------------------
+
+class InvalidUsage(Exception):
+    status_code = 400
+
+    def __init__(self, message, status_code=None, payload=None):
+        Exception.__init__(self)
+        self.message = message
+        if status_code is not None:
+            self.status_code = status_code
+        self.payload = payload
+
+    def to_dict(self):
+        http_resp = dict(self.payload or ())
+        http_resp['message'] = self.message
+        return http_resp
+
+@app.errorhandler(InvalidUsage)
+def handle_invalid_usage(error):
+    response = jsonify(error.to_dict())
+    response.status_code = error.status_code
+    return response
 #---------------------Rest API---------------------
 class Node(Resource):
-    def post(self):
-        if request.form['_method'] == 'put':
-            return self.put()
-        else:
-            # g.db.execute('insert into nodes (contents, renderer) values (?, ?)',
-            #             [request.form['contents'], request.form['renderer']])
-            # g.db.commit()
-            return "FOO" #jsonify(message='New node was successfully created')
+    def post(self, node_id):
+        g.db.execute('''UPDATE nodes 
+                        SET contents=(?),renderer=(?) 
+                        WHERE id=(?)''',
+                        [request.form['contents'], request.form['renderer'], node_id])
+        g.db.commit()
+        g.db.commit()
+        return jsonify(message='Node was successfully updated.', id='1')
+
     def put(self):
-        g.db.execute('insert into nodes (contents, renderer) values (?, ?)',
+        """
+        This inserts a new node into the database, with the node_id
+        automatically assigned
+        """
+        g.db.execute('INSERT INTO nodes (contents, renderer) VALUES(?, ?)',
                      [request.form['contents'], request.form['renderer']])
         g.db.commit()
-        return jsonify(message='New node was successfully created')
+        return jsonify(message='New node was successfully created', id='1')
 
-class Tree(Resource):
-    def get(self):
-        return {"nodes":3}
+    # takes a node_id
+    # returns a single node
+    # if node_id out of range, returns null
     def get(self, node_id):
-        cur = g.db.execute("SELECT * FROM nodes WHERE node_id=(?) " , node_id)
-        rv = cur.fetchall()[0]
-        cur = g.db.execute("SELECT name, dest FROM links WHERE origin=(?) ", node_id)
-        children = {}
-        for k,val in enumerate(cur.fetchall()):
-            children[val['name']] = val['dest']
-        rv['children'] = children
-        cur.close()
-        return rv
+        cursor = g.db.execute('''SELECT n.id, n.contents, n.renderer, c.children 
+                              FROM nodes AS n LEFT JOIN children AS c 
+                              ON n.id = c.parent_id WHERE id=(?)''', node_id)
+        return cursor.fetchone()
+
+class Children(Resource):
+    def put(self):
```

Comment:
This API is written differently I think. Adding a child relationship means you specify the parent's ID in the data payload as well as the IDs of the children